### PR TITLE
Add support for streaming

### LINKED ISSUE #12: Support streaming on the chat interface

### PR DESCRIPTION
**Pull request by Nintorac**
Closes #12

Should be good to go but haven't tested from the ground up in a clean environment

### CODE REVIEW

File: src/simple_ai/server.py [hunk 1]
```diff
@@ -80,8 +145,18 @@ async def chat_complete(body: Annotated[ChatCompletionInput, Body(example=dummy_
...
-    output = format_chat_response(model_name=llm.name, predictions=predictions)
-    return output
+
+    background_tasks.add_task(lambda f: f.close(), predictions_stream)
+
+    # predictions_stream = [[{'role': f"assistnat "}]] + [[{'content': f"{i} "}] for i in range(10)]
+    uuid = uuid4().hex
+    current_timestamp = int(dt.now().timestamp())
+    postprocessed = map(
+        partial(format_chat_delta_response, current_timestamp, uuid, body.model), predictions_stream
```
Comment (lhenault) [inline src/simple_ai/server.py:155]: Looks like every part of the stream will share the same timestamp and id, shouldn't it be different for each?

**Comment by lhenault** [comment]
My main complain is that you've stolen my next weekend project, but very cool job. Thanks  :)

> haven't tested from the ground up in a clean environment

Will try this and let you know if I see any issue arising.

**Comment by lhenault** [comment]
I've tried with a fresh install and some dummy model server. It works seamlessly on both `chat` and `complete`, in streaming and non streaming mode. Well done! Once the minor comments are addressed I'd be happy to merge this.

File: src/simple_ai/server.py (hunk 1 above)
Comment (Nintorac) [reply src/simple_ai/server.py:155]: yea, this matches openAI 

```
(base) ➜  ~ curl https://api.openai.com/v1/completions \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -H "Authorization: Bearer $OPENAI_API_KEY" \
  -d '{
    "model": "davinci",
    "prompt": "some kind of text",
    "max_tokens": 1000,
    "stream": true
}'
data: {"id": "cmpl-74SD2xoWGle0UNOhGPDZO2WMc1Mes", "object": "text_completion", "created": 1681296104, "choices": [{"text": " of", "index": 0, "logprobs": null, "finish_reason": null}], "model": "davinci"}

data: {"id": "cmpl-74SD2xoWGle0UNOhGPDZO2WMc1Mes", "object": "text_completion", "created": 1681296104, "choices": [{"text": " cl", "index": 0, "logprobs": null, "finish_reason": null}], "model": "davinci"}

data: {"id": "cmpl-74SD2xoWGle0UNOhGPDZO2WMc1Mes", "object": "text_completion", "created": 1681296104, "choices": [{"text": "otted", "index": 0, "logprobs": null, "finish_reason": null}], "model": "davinci"}

data: {"id": "cmpl-74SD2xoWGle0UNOhGPDZO2WMc1Mes", "object": "text_completion", "created": 1681296104, "choices": [{"text": ",", "index": 0, "logprobs": null, "finish_reason": null}], "model": "davinci"}

^C

```

**Comment by Nintorac** [comment]
heh nearly stopped myself at 1! Ended up doing complete as well for streaming langchain but it didn't stream the response to console :-1:  

I also did a few experiment with the Open AI Evals repo, rwkv doesn't do well there unfortunately but cool that SimpleAI works with their tools!

**Comment by lhenault** [comment]
> heh nearly stopped myself at 1!

Not sure I got this part :)

> streaming langchain but it didn't stream the response to console 👎

Is that because of the issue #18 or something else?

> I also did a few experiment with the Open AI Evals repo, rwkv doesn't do well there unfortunately but cool that SimpleAI works with their tools!

Yeah that's one of the goals of the project, having an easy way to compare and perhaps switch to alternative models. Cool to know it worked on that repo! If not `rmkv`, we should get competing open alternatives soon, one can hope. :)

LGTM. Feels ready to merge, except if I've missed something here.

**Comment by Nintorac** [comment]
Nearly stopped at just chat streaming, instead of also completion streaming. No I found a workaround for that by editing the langchain module files, langchain just collects all the responses and prints to console in a single go rather than streaming. 

Would be really interested in a 100b rwkv but I think that's a whole next level of engineering and $$. OpenAssistant are pushing the right buttons to make it achievable though

Yea I think it's ready, have been using it without bugs for a few days at least :)

**Pull request merged**